### PR TITLE
Implement Smart will update functionality

### LINKED ISSUE #117: [CONTRACT] Implement Smart will update functionality
## Description 📝

Implement a Smart Will management system in the AccountData component that allows members to designate another member's address as their will address. This feature provides inheritance functionality where members can assign beneficiaries. The system includes validation mechanisms, storage management, and getter functions for will addresses and durations.

## Proposed Steps 📝

1. **Define Data Structures and Constants**
   - Define `DEFAULT_WILL_DURATION` constant (90 days in seconds: 7,776,000)
   - Create error types for validation failures
   - Define events for will creation and updates

2. **Implement Storage Maps**
   - Create storage maps for will management:
     - `smart_will_to_member: Map<ContractAddress, ContractAddress>` (will address → member address)
     - `member_to_smart_will: Map<ContractAddress, ContractAddress>` (member address → will address)
     - `member_to_will_duration: Map<ContractAddress, u64>` (member address → duration in seconds)
     - `member_will_creation_time: Map<ContractAddress, u64>` (member address → timestamp when will was set)

3. **Core Function Implementation**
   - Implement `update_smart_will(will_address: ContractAddress)` function:
     - Validate caller is a member
     - Validate will_address is not a member
     - Validate will_address does not belongs to another member
     - Check that member does not have a will wallet or if duration has not elapsed
     - Update storage maps accordingly
     - Set default duration if first-time setup

4. **Validation Logic**
   - Create helper function `validate_will_conditions(member: ContractAddress, will_address: ContractAddress)`:
     - Check if will_address is not in members list
     - Verify will_address does not belongs to an existing member
     - Check existing will duration constraints
   - Implement `can_update_will(member: ContractAddress) -> bool`:
     - Return true if no existing will OR duration has not elapsed

5. **Getter Functions**
   - Implement `get_member_will_address(member: ContractAddress) -> ContractAddress`:
     - Return the will address for a given member
     - Return zero address if no will is set
   - Implement `get_member_will_duration(member: ContractAddress) -> u64`:
     - Return the will duration for a given member
     - Return 0 if no will is set
   - Implement `get_remaining_will_time(member: ContractAddress) -> u64`:
     - Calculate and return remaining time before will can be updated
     - Return 0 if will can be updated immediately

6. **Integration and Events**
   - Emit events for will creation and updates
   - Integrate with existing member management system
   - Add access control modifiers where necessary

7. **Testing Implementation**
   - Write comprehensive unit tests for all validation conditions
   - Test edge cases
   - Test storage state consistency across updates
   - Performance testing for large numbers of will assignments

## Acceptance Criteria ✅

- [ ] `update_smart_will` function is implemented with all validation conditions:
  - [ ] Validates caller is a member
  - [ ] Ensures will_address is not a member
  - [ ] Confirms will_address does not belong to another member
  - [ ] Checks will duration constraints

- [ ] Storage maps are properly implemented:
  - [ ] `smart_will_to_member` mapping
  - [ ] `member_to_smart_will` mapping  
  - [ ] `member_to_will_duration` mapping
  - [ ] `member_will_creation_time` mapping

- [ ] Default 90-day duration is set for new will addresses
- [ ] Existing will addresses can only be updated before duration elapses
- [ ] Getter functions work correctly:
  - [ ] `get_member_will_address` returns correct address or zero
  - [ ] `get_member_will_duration` returns correct duration or zero
  - [ ] `get_remaining_will_time` calculates time correctly

- [ ] Proper error handling and validation:
  - [ ] Appropriate errors for invalid conditions
  - [ ] Protection against unauthorized updates

- [ ] Events are emitted for all will-related actions

## Additional Context 📌
- The 90-day default (7,776,000 seconds) prevents frequent will changes while allowing reasonable updates
- Use `get_block_timestamp()` for consistent blockchain time
- Changes should be made in the AccountData component.

### PR DESCRIPTION
**Pull request by hakymulla**
## Description 📝
[CONTRACT] Implement Smart will update functionality

## Related Issues 🔗
Closes #117 

## Changes Made 🚀
- [x] ✨ Feature Implementation 
- [ ] 🐛 Bug Fix 
- [ ] 📚 Documentation Update 
- [ ] 🔨 Refactoring 
- [ ] ❓ Others (Specify) 

## Checklist ✅
- [x] 🛠 I have tested these changes. 
- [ ] 📖 I have updated the documentation (if applicable). 
- [x] 🎨 This PR follows the project's coding style. 
- [x] 🧪 I have added necessary tests (if applicable). 

## Summary by CodeRabbit

* **New Features**
  * Introduced a "Smart Will" feature, allowing members to assign, update, and manage will addresses with configurable durations and cooldown periods.
  * Added new error messages for improved feedback on will-related actions.
  * Provided new interface methods for accessing and managing smart will data.

* **Tests**
  * Added comprehensive tests to validate smart will functionality, including will assignment, duration enforcement, access control, and error handling.

* **Chores**
  * Added a coding guideline document to support best practices and workflows for Cairo development.

### CODE REVIEW

File: src/account_data.cairo [hunk 1]
```diff
@@ -652,5 +762,32 @@ pub mod AccountData {
...
+                Errors::ERR_WILL_ADDRESS_ALREADY_ASSIGNED
+            );
+
+            // Check if member can update their will
+            let creation_time = self.member_will_creation_time.entry(member).read();
+            if creation_time != 0 {
+                let duration = self.member_to_will_duration.entry(member).read();
+                let current_time = get_block_timestamp();
+                let elapsed_time = current_time - creation_time;
+                assert(elapsed_time >= duration, Errors::ERR_WILL_DURATION_NOT_ELAPSED);
```
Comment (codebestia) [inline src/account_data.cairo:789]:
```suggestion
                assert(duration > current_time, Errors::ERR_WILL_DURATION_NOT_ELAPSED);
```

**Review by codebestia — CHANGES_REQUESTED** [review]
Good Job @hakymulla 
Just some minor fixes and we are good to go.

**Review by codebestia — APPROVED** [review]
LGTM!
Thank you for your contribution.
